toolkit/tools/pkg/imagecustomizerlib/imagecustomizer.go

```diff
@@ -71,124 +102,252 @@ func CustomizeImageWithConfigFile(buildDir string, configFile string, imageFile
+) error {
+	err := validateConfig(baseConfigPath, config, rpmsSources, useBaseImageRpmRepos)
 	if err != nil {
-		return err
+		return fmt.Errorf("invalid image config:\n%w", err)
 	}
 
-	// Convert image file to raw format, so that a kernel loop device can be used to make changes to the image.
-	rawImageFile := filepath.Join(buildDirAbs, BaseImageName)
+	ic, err := createImageCustomizer(buildDir, imageFile,
+		baseConfigPath, config,
+		useBaseImageRpmRepos, rpmsSources, enableShrinkFilesystems, outputSplitPartitionsFormat,
+		outputImageFormat, outputImageFile)
+	if err != nil {
+		return fmt.Errorf("failed to create image customizer:\n%w", err)
+	}
 	defer func() {
-		cleanupErr := file.RemoveFileIfExists(rawImageFile)
+		cleanupErr := ic.cleanUp()
 		if cleanupErr != nil {
 			if err != nil {
-				err = fmt.Errorf("%w:\nfailed to clean-up (%s): %w", err, rawImageFile, cleanupErr)
+				err = fmt.Errorf("%w:\nfailed to clean-up:\n%w", err, cleanupErr)
 			} else {
-				err = fmt.Errorf("failed to clean-up (%s): %w", rawImageFile, cleanupErr)
+				err = fmt.Errorf("failed to clean-up:\n%w", cleanupErr)
 			}
 		}
 	}()
 
-	logger.Log.Infof("Creating raw base image: %s", rawImageFile)
-	err = shell.ExecuteLiveWithErr(1, "qemu-img", "convert", "-O", "raw", imageFile, rawImageFile)
+	err = ic.convertInputImageToWriteableFormat()
 	if err != nil {
-		return fmt.Errorf("failed to convert image file to raw format:\n%w", err)
+		return fmt.Errorf("failed to convert input image to a raw image:\n%w", err)
 	}
 
+	if ic.customizeOSPartitions {
+		err = ic.customizeOSContents()
+		if err != nil {
+			return fmt.Errorf("failed to customize raw image:\n%w", err)
+		}
+	} else {
+		logger.Log.Infof("Skipping customizing OS partitions because none is specified")
+	}
+
+	err = ic.convertWriteableFormatToOutputImage()
+	if err != nil {
+		return fmt.Errorf("failed to convert customized raw image to output format:\n%w", err)
+	}
+
+	logger.Log.Infof("Success!")
+
+	return nil
+}
+
+func (ic *ImageCustomizer) convertInputImageToWriteableFormat() error {
+
+	logger.Log.Infof("Converting input image to a writeable format")
+
+	if filepath.Ext(ic.inputImageFile) == ".iso" {
+
+		isoBuilder, err := createIsoBuilderFromIsoImage(ic.buildDir, ic.buildDirAbs, ic.inputImageFile)
+		if err != nil {
+			if ic.isoBuilder != nil {
+				ic.isoBuilder.cleanUp()
+			}
+			return fmt.Errorf("failed to load input iso artifacts:\n%w", err)
+		}
+		ic.isoBuilder = isoBuilder
+
+		// If the input is a LiveOS iso and there are OS customizations are
+		// defined, we create a writeable disk image so that mic can modify
+		// it. If no OS customizations are defined, we can skip this step and
+		// just re-use the existing squashfs.
+		if ic.customizeOSPartitions {
+			err := ic.isoBuilder.createWriteableImageFromSquashfs(ic.buildDir, ic.rawImageFile)
+			if err != nil {
+				return fmt.Errorf("failed to create writeable image:\n%w", err)
+			}
```

Comment:
Should `ic.isoBuilder` be deleted here? When `customizeOSPartitions` is `true`, `ic.isoBuilder` isn't used after this point.

Reply:
It actually needs to survive post this point...
isoBuilder has populated its internal state with all the relevant artifacts. That state can be used to regenerate the iso again without re-building any of these artifacts.
I've moved isoBuild out of the parameter structure so that it's not a side effect anymore. It's now returned from `convertInputImageToWriteableFormat` and is fed into `convertWriteableFormatToOutputImage`.

Reply:
But you create a new copy of the `LiveOSIsoBuilder` type in the `createLiveOSIsoImage` function. So, surely the resources being held by `ic.isoBuilder` aren't needed?

Reply:
I see you're point. You're right. I think we can make that decision right here.
pending.

Reply:
I've found a flow where we do need the input iso builder object even when we're re-creating the squashfs. So, I need to keep it around.
Basically, when doing iso->iso creation, even if we re-create the squashfs (boot + rootfs), we still need the original iso state to determine what extra files the user has packed into it (in a previous customization), and copy them unto the new iso as well.